classes/OJSSwordDeposit.inc.php

```diff
@@ -0,0 +1,202 @@
+
+		// The article can be published or not. Support either.
+		if (is_a($this->_article, 'PublishedArticle')) {
+			$doi = $this->_article->getStoredPubId('doi');
+			if ($doi !== null) $this->_package->setIdentifier($doi);
+		}
+
+		foreach ($this->_article->getAuthors() as $author) {
+			$creator = $author->getFullName(true);
+			$affiliation = $author->getLocalizedAffiliation();
+			if (!empty($affiliation)) $creator .= "; $affiliation";
+			$this->_package->addCreator($creator);
+		}
+
+// 		TODO check with Nate why cistation style plugin is not enabled by default
+// 		also which citation style is DSpace expecting?
+// 		if (is_a($this->_article, 'PublishedArticle')) {
+// 			$plugin = PluginRegistry::getPlugin('generic', 'citationstylelanguageplugin');
+// 			if ($plugin) {
+// 				$citation = $plugin->getCitation($request, $this->_article, 'bibtex');
+// 				$this->_package->setCitation(html_entity_decode(strip_tags($citation), ENT_QUOTES, 'UTF-8'));
+// 			}
+// 		}
+	}
+
+	/**
+	 * Add a file to a package. Used internally.
+	 * @param $submissionFile SubmissionFile
+	 */
+	public function _addFile($submissionFile) {
+		$targetFilename = $this->_outPath . '/files/' . $submissionFile->getLocalizedName();
+		copy($submissionFile->getFilePath(), $targetFilename);
+		$this->_package->addFile($submissionFile->getLocalizedName(), $submissionFile->getFileType());
+	}
+
+	/**
+	 * Add all article galleys to the deposit package.
+	 */
+	public function addGalleys() {
+		foreach ($this->_article->getGalleys() as $galley) {
+			$this->_addFile($galley->getFile());
+		}
+	}
+
+	/**
+	 * Add the single most recent editorial file to the deposit package.
+	 * @return boolean true iff a file was successfully added to the package
+	 */
+	public function addEditorial() {
+		$submissionFileDao = DAORegistry::getDAO('SubmissionFileDAO');
+		$submissionFiles = $submissionFileDao->getBySubmissionId($this->_article->getId());
+		// getBySubmission orders results by id ASC, let's reverse the array to have recent files first
+		$submissionFiles = array_reverse($submissionFiles, true);
+		$files = array();
+		foreach ($submissionFiles as $submissionFile) {
+			$fileStage = $submissionFile->getFileStage();
+			if (!isset($files[$fileStage])) {
+				$files[$fileStage] = array();
+			}
+			$files[$fileStage][] = $submissionFile;
+		}
+		// Move through stages in reverse order and try to use them.
+		$stages = array(
+			SUBMISSION_FILE_PRODUCTION_READY,
+			SUBMISSION_FILE_COPYEDIT,
+			SUBMISSION_FILE_REVIEW_FILE,
+			SUBMISSION_FILE_SUBMISSION
+		);
+		$mostRecentEditorialFile = null;
+		foreach ($stages as $subFileStage) {
+			if (isset($files[$subFileStage])) {
+				$mostRecentEditorialFile = array_shift($files[$subFileStage]);
+				$this->_addFile($mostRecentEditorialFile);
+				return true;
+			}
+		}
+		// TODO when the above didn't work, SWORD plugin for OJS2 was trying "the editor version"
```

Comment:
The above code should walk backwards from the latest to the oldest in-progress version by file stage, so effectively it already does what OJS 2.x used to do (with OJS 2.x's more limited conception of a workflow).

Reply:
👍